### DISCUSSION TITLE
Running the Breakout Example

### DISCUSSION THREAD

**Discussion by guinness-1** [bevyengine/bevy#2791] [Q&A] [answered]
Hello,

I'm just trying Bevy for the first time and am struggling to run the example programs from this repo. I've seen some advice about editing Xorg conf files, but nothing has worked so far. I'm on a Ubuntu 20.04 VM. Let me know what other information I can provide. Below is my output from running the example. Thanks for any advice!
```
$ cargo run --example breakout
    Finished dev [unoptimized + debuginfo] target(s) in 0.63s
     Running `target/debug/examples/breakout`
WARNING: lavapipe is not a conformant vulkan implementation, testing use only.
vulkan: No DRI3 support detected - required for presentation
Note: you can probably enable DRI3 in your Xorg config
thread 'main' panicked at 'Error in Device::create_swap_chain: surface does not support the adapter's queue family', /home/jfloyd/.cargo/registry/src/github.com-1ecc6299db9ec823/wgpu-0.7.1/src/backend/direct.rs:129:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
Sep 08 10:20:10.302 ERROR gpu_descriptor::allocator: `DescriptorAllocator` is dropped while some descriptor sets were not deallocated
Sep 08 10:20:10.302 ERROR gpu_alloc::block: Memory block wasn't deallocated
Sep 08 10:20:10.302 ERROR gpu_alloc::block: Memory block wasn't deallocated
```

**Comment by Nilirad** [comment, marked as answer]
I'm not on Linux on the majority of my time, but it looks like you're using mesa vulkan driver. Chances are you're using an Nvidia GPU. Probably the issue will disappear if you switch to the proprietary driver. Don't take for granted my word though. Wait for better support if you don't know how to swap drivers.

**Reply by guinness-1** [reply]
Thanks @Nilirad. My host machine definitely has an nvidia GPU, but not sure how that translates to inside the VM. There, I see the following GPU info:

```
$ hwinfo --gfxcard --short
graphics card:                                                  
                       VMware VMWARE0405
```

**Reply by minecrawler** [reply]
Looks like you are using the software GPU. That might be [a problem](https://www.reddit.com/r/vulkan/comments/j71x86/vulkan_support_in_virtual_machines/)

If you want to use your discreet GPU inside a VM, you will need to pass through dedicated access to the hardware to the VM, so that the host computer will no longer be able to use it. Most hypervisors offer such an option.

I recommend doing so when you either have GPUs to spare on your desktop machine, or if you are running the hypervisor on a server. Otherwise, the VMWare graphics card is a software implementation with only limited features and performance.

For VMWare, there is a [help article](https://docs.vmware.com/en/VMware-Workstation-Player-for-Linux/16.0/com.vmware.player.linux.using.doc/GUID-DB7C4F11-1588-4DD3-BB1F-FBABFA570E3A.html) available to get you started and several sources to configure it, depending on what product you use. While I have no experience with your setup, I guess that Vulkan is not supported. You will likely have to use OpenGL for that VM in that configuration...

**Comment by guinness-1** [comment]
Oh, I'm happy to use OpenGL to get started. Does Bevy have an option for that?

**Reply by bjorn3** [reply]
Not yet. The latest version of wgpu has support for it I believe, but we will need to wait with updating until the 2021 edition gets released as the latest wgpu version requires `resolver = "2"` in the `Cargo.toml` of the game itself. With the 2021 edition this will be the default.

**Reply by guinness-1** [reply]
Alright, thanks all. I installed on baremetal and everything works fine. It would be nice if VMs were a more feasible option, but I'm all set up to go explore Bevy for now. It looks great!
